### PR TITLE
feat: engine type implementation (will replace node api preview feature)

### PR DESCRIPTION
**Pull request by williamluke4**
Both the CLI and Client are still by default using the `binary` but this allows for the change over to be much easier.  
#### ENV OVERRIDE
```env
PRISMA_CLIENT_ENGINE_TYPE="library" | "binary"
```
### Schema
```groovy
generator client {
	provider    = "prisma-client-js"
	engineType  = "binary" | "library" 
}
```
#### ENV OVERRIDE
```env
PRISMA_CLI_QUERY_ENGINE_TYPE="binary" | "library" 
```

### CODE REVIEW

File: src/packages/client/src/generation/generateClient.ts [hunk 1]
```diff
@@ -234,14 +236,17 @@ export async function generateClient({
       })
     }
   }
-  const enginePath = useNodeAPI
-    ? binaryPaths.libqueryEngine
-    : binaryPaths.queryEngine
+  const binaryPath =
```
Comment (williamluke4) [inline src/packages/client/src/generation/generateClient.ts:239]: revert this change

File: src/packages/client/src/generation/generateClient.ts [hunk 2]
```diff
@@ -280,9 +285,10 @@ export async function generateClient({
         await copyFile(filePath, target)
         continue
       }
-      const binaryName = useNodeAPI
-        ? BinaryType.libqueryEngine
-        : BinaryType.queryEngine
+      const binaryName =
```
Comment (janpio) [inline src/packages/client/src/generation/generateClient.ts:288]: => engineName?

File: packages/cli/src/Version.ts [hunk 3]
```diff
@@ -84,7 +83,11 @@ export class Version implements Command {
       ['@prisma/client', prismaClientVersion ?? 'Not found'],
       ['Current platform', platform],
       [
-        `Query Engine${useNodeAPI ? ' (Node-API)' : ''}`,
+        `Query Engine${
+          cliQueryEngineBinaryType === BinaryType.libqueryEngine
+            ? ' (Node-API)'
+            : ''
+        }`,
```
Comment (Jolg42) [inline packages/cli/src/Version.ts:90]:
```suggestion
        `Query Engine${
          cliQueryEngineBinaryType === BinaryType.libqueryEngine
            ? ' (Node-API)'
            : ' (Binary)'
        }`,
```
So it's clearer maybe?

File: packages/client/src/__tests__/integration/happy/chmod/test.ts [hunk 4]
```diff
@@ -1,13 +1,17 @@
...
+  getClientEngineType,
+} from '../../../../runtime/utils/getClientEngineType'
 import { generateTestClient } from '../../../../utils/getTestClient'
 
 // Tests that no error is being thrown when the binary is manually set to chmod 644 because Client fixes that itself
 test('chmod', async () => {
   await generateTestClient()
   const platform = await getPlatform()
-  if (!process.env.PRISMA_FORCE_NAPI) {
+  if (!(getClientEngineType() === ClientEngineType.NodeAPI)) {
```
Comment (Jolg42) [inline packages/client/src/__tests__/integration/happy/chmod/test.ts:14]: A bit weird written like this I think, why not like this?
```suggestion
  if (getClientEngineType() !== ClientEngineType.NodeAPI) {
```

File: packages/client/src/__tests__/integration/happy/client-engine/test.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    ClientEngineType.Binary,
+    ClientEngineType.NodeAPI,
+    undefined,
+  ]
+  const envVars = {
+    PRISMA_CLIENT_ENGINE_TYPE: engineTypes,
+  }
+  engineTypes.forEach((engineType) => {
+    for (const envVar in envVars) {
+      envVars[envVar].forEach((value) => {
```
Comment (Jolg42) [inline packages/client/src/__tests__/integration/happy/client-engine/test.ts:70]: Would be nice to use for in everywhere I think? Would be a bit easier to read

File: packages/client/src/__tests__/integration/happy/client-engine/test.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+          engineType,
+          envVar,
+          value,
+        )
+        test(`expects(${expectedClientEngine}) | ${envVar}=${value} | engineType=${engineType}`, async () => {
+          expect.assertions(2)
+          const schema = buildSchema(engineType)
+
+          // Setup Project in tmp dir
+          const projectDir = fs.mkdtempSync(`${os.tmpdir()}${path.sep}`)
```
Comment (Jolg42) [inline packages/client/src/__tests__/integration/happy/client-engine/test.ts:81]: Would be nice to delete the temp dir after to cleanup

File: packages/client/src/runtime/utils/getClientEngineType.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { GeneratorConfig } from '@prisma/generator-helper'
+
+export enum ClientEngineType {
+  NodeAPI = 'node-api',
+  Binary = 'binary',
+}
+export const DEFAULT_CLIENT_ENGINE_TYPE = ClientEngineType.Binary
+
+export function getClientEngineType(
+  generator?: GeneratorConfig,
```
Comment (Jolg42) [inline packages/client/src/runtime/utils/getClientEngineType.ts:10]:
```suggestion
  generatorConfig?: GeneratorConfig,
```
Just a thought to make it easier to know what it expects

File: packages/client/src/runtime/utils/getClientEngineType.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+function getEngineTypeFromEnvVar() {
+  const engineType = process.env.PRISMA_CLIENT_ENGINE_TYPE
+  if (engineType) {
+    if (engineType === ClientEngineType.NodeAPI) {
+      return ClientEngineType.NodeAPI
+    } else if (engineType === ClientEngineType.Binary) {
+      return ClientEngineType.Binary
+    }
+  }
```
Comment (Jolg42) [inline packages/client/src/runtime/utils/getClientEngineType.ts:33]: could be simplified to 
```suggestion
if (engineType === ClientEngineType.NodeAPI) {
  return ClientEngineType.NodeAPI
} else if (engineType === ClientEngineType.Binary) {
  return ClientEngineType.Binary
}
```

File: packages/client/src/runtime/utils/getClientEngineType.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  return undefined
```
Comment (Jolg42) [inline packages/client/src/runtime/utils/getClientEngineType.ts:34]: Could be "simplified" with

```suggestion
if (engineType === ClientEngineType.NodeAPI) {
  return ClientEngineType.NodeAPI
} else if (engineType === ClientEngineType.Binary) {
  return ClientEngineType.Binary
} else {
  return undefined
}
```

File: packages/client/src/utils/ensureTestClientQueryEngine.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  const queryEngineLibraryPath = path.join(
+    enginesPath,
+    getNodeAPIName(platform, 'fs'),
+  )
+  const queryEngineBinaryPath = path.join(
+    enginesPath,
+    `query-engine-${platform}${platform === 'windows' ? '.exe' : ''}`,
+  )
+
+  // TMP
```
Comment (Jolg42) [inline packages/client/src/utils/ensureTestClientQueryEngine.ts:28]: Curious about what is this TMP comment :)

File: scripts/ci/publish.ts [hunk 11]
```diff
@@ -858,32 +858,21 @@ async function testPackages(
...
-    order = [
-      '@prisma/sdk',
-      '@prisma/migrate',
-      '@prisma/client',
-      'prisma',
-      '@prisma/integration-tests',
-    ]
+    // This is a tmp workaround
+    console.log('SKIPPING')
+    order = []
```
Comment (Jolg42) [inline scripts/ci/publish.ts:875]: Does that mean that this needs to be removed when we merge?

**Review by Jolg42 — APPROVED** [review]
Looks good (see comments)

Question: Is this validated by VSCode / engine validation?
`engineType  = "binary" | "library"`

File: .buildkite/test/run.sh [hunk 12]
```diff
@@ -41,20 +41,32 @@ function retry {
   return 0
 }
 
-# Only for job 2 = Node-API
-if [ "$BUILDKITE_PARALLEL_JOB" = "2" ]; then
-  export PRISMA_FORCE_NAPI=true
+# The below is required as during install required engines are download, to this makes sure the the engines being tested are already present
```
Comment (janpio) [inline .buildkite/test/run.sh:44]:
```suggestion
# The below is required as during install required engines are downloaded, so this makes sure the engines being tested are already present 
```

File: .buildkite/test/run.sh [hunk 13]
```diff
@@ -41,20 +41,32 @@ function retry {
   return 0
 }
 
-# Only for job 2 = Node-API
-if [ "$BUILDKITE_PARALLEL_JOB" = "2" ]; then
-  export PRISMA_FORCE_NAPI=true
```
Comment (janpio) [inline .buildkite/test/run.sh:46]: Nothing necessary for job 2 any more?

File: packages/client/fixtures/blog/prisma/schema.prisma [hunk 14]
```diff
@@ -5,7 +5,8 @@ datasource db {
 
 generator client {
   provider        = "prisma-client-js"
-  previewFeatures = ["napi", "selectRelationCount"]
+  previewFeatures = ["selectRelationCount"]
+  engineType = "binary"
```
Comment (janpio) [inline packages/client/fixtures/blog/prisma/schema.prisma:9]: is this necessary here?

File: packages/client/src/__tests__/generation/generator.test.ts [hunk 15]
```diff
@@ -43,7 +47,7 @@ describe('generator', () => {
     }
     manifest.requiresEngineVersion = 'ENGINE_VERSION_TEST'
 
-    if (process.env.PRISMA_FORCE_NAPI) {
+    if (getClientEngineType(generator.config) === ClientEngineType.NodeAPI) {
```
Comment (janpio) [inline packages/client/src/__tests__/generation/generator.test.ts:50]: Why does this get a param here, while not in many other of the usages of it?

File: packages/client/src/utils/ensureTestClientQueryEngine.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  if (
+    clientEngineType === ClientEngineType.NodeAPI &&
```
Comment (millsp) [inline packages/client/src/utils/ensureTestClientQueryEngine.ts:30]:
```
ClientEngineType.Library
```

File: scripts/ci/publish.ts [hunk 17]
```diff
@@ -893,10 +882,16 @@ async function testPackages(
     const pkg = packages[pkgName]
     if (pkg.packageJson.scripts.test) {
       console.log(`\nTesting ${chalk.magentaBright(pkg.name)}`)
-      if (process.env.BUILDKITE_PARALLEL_JOB === '2') {
+      // Sets ENV to override engines
+      if (process.env.BUILDKITE_PARALLEL_JOB === '0') {
```
Comment (millsp) [inline scripts/ci/publish.ts:886]: Why did we usually test all packages but not the client in step 0 (previously)?

File: packages/client/src/generation/utils.ts [hunk 18]
```diff
@@ -366,25 +367,26 @@ export function unique<T>(arr: T[]): T[] {
...
-          )
-          .join('\n')
-      : ''
-  }`
+  const getQueryEngineFilename = (p: Platform) =>
+    clientEngineType === ClientEngineType.Binary
+      ? `query-engine-${p}`
+      : getNodeAPIName(p, 'fs')
+
+  const buildAnnotation = (p: Platform) => {
```
Comment (millsp) [inline packages/client/src/generation/utils.ts:383]: Nice

File: packages/client/src/__tests__/generation/generator.test.ts (hunk 15 above)
Comment (millsp) [reply packages/client/src/__tests__/generation/generator.test.ts:50]: I think that this is not `getCliEngineType`. The one for the `client` looks into the generator

File: .github/workflows/test.yml [hunk 19]
```diff
@@ -116,17 +116,16 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        engine: ['napi', 'binary']
+        engineType: ['node-api', 'binary']
```
Comment (millsp) [inline .github/workflows/test.yml:119]: Could we rename this to `library` ?

File: packages/client/fixtures/blog/prisma/schema.prisma (hunk 14 above)
Comment (williamluke4) [reply packages/client/fixtures/blog/prisma/schema.prisma:9]: No this is just a fixture used for development

File: packages/cli/src/Version.ts (hunk 3 above)
Comment (williamluke4) [reply packages/cli/src/Version.ts:90]: Implemented

File: packages/client/src/__tests__/integration/happy/client-engine/test.ts (hunk 5 above)
Comment (williamluke4) [reply packages/client/src/__tests__/integration/happy/client-engine/test.ts:70]: `for in` is mainly for objects changed them to `for of`

File: packages/client/src/utils/ensureTestClientQueryEngine.ts (hunk 10 above)
Comment (williamluke4) [reply packages/client/src/utils/ensureTestClientQueryEngine.ts:28]: Removed the comment

File: packages/client/src/__tests__/generation/generator.test.ts (hunk 15 above)
Comment (williamluke4) [reply packages/client/src/__tests__/generation/generator.test.ts:50]: Removed as it is not necessary here.

File: packages/client/src/utils/ensureTestClientQueryEngine.ts (hunk 16 above)
Comment (williamluke4) [reply packages/client/src/utils/ensureTestClientQueryEngine.ts:30]: Implemented

File: .github/workflows/test.yml (hunk 19 above)
Comment (williamluke4) [reply .github/workflows/test.yml:119]: Implemented

File: scripts/ci/publish.ts (hunk 17 above)
Comment (williamluke4) [reply scripts/ci/publish.ts:886]: Think it was an attempt to speed up the buildkite tests

File: .buildkite/test/run.sh (hunk 13 above)
Comment (williamluke4) [reply .buildkite/test/run.sh:46]: Nope
Job 0 Tests all with the Node-API Library
Job 1 Tests all with the Binary

File: scripts/ci/publish.ts (hunk 11 above)
Comment (williamluke4) [reply scripts/ci/publish.ts]: I have removed this now that I know how to change the [buildkite parallelism](https://github.com/prisma/prisma/blob/main/.buildkite/test/test.yml#L3)

File: .buildkite/test/run.sh (hunk 13 above)
Comment (janpio) [reply .buildkite/test/run.sh:46]: Nice.

File: packages/client/fixtures/blog/prisma/schema.prisma (hunk 14 above)
Comment (janpio) [reply packages/client/fixtures/blog/prisma/schema.prisma:9]: So not a test? Might be useful to rename this then from `blog` to `blog-used-during-development` or something like that.

File: scripts/ci/publish.ts (hunk 17 above)
Comment (janpio) [reply scripts/ci/publish.ts:882]: I think we _extracted_ the client into its own job (hence we had job 0, 1 and 2) to speed things up indeed.

File: packages/client/src/__tests__/generation/generator.test.ts [hunk 20]
```diff
@@ -214,7 +218,7 @@ describe('generator', () => {
     }
     manifest.requiresEngineVersion = 'ENGINE_VERSION_TEST'
 
-    if (process.env.PRISMA_FORCE_NAPI) {
+    if (getClientEngineType(generator.config) === ClientEngineType.Library) {
```
Comment (janpio) [inline packages/client/src/__tests__/generation/generator.test.ts:221]:
```suggestion
    if (getClientEngineType() === ClientEngineType.Library) {
```
I think.

File: packages/client/src/__tests__/integration/errors/missing-binary/test.ts [hunk 21]
```diff
@@ -10,17 +14,18 @@ test('missing-binary', async () => {
...
+      ? path.join(
+          __dirname,
+          'node_modules/.prisma/client',
+          getNodeAPIName(platform, 'fs'),
+        )
+      : path.join(
+          __dirname,
+          'node_modules/.prisma/client',
+          `query-engine-${platform}`,
+        )
```
Comment (janpio) [inline packages/client/src/__tests__/integration/errors/missing-binary/test.ts:28]: Should we maybe move this to somewhere else and just import it here vs. re-create this in each test?

File: packages/client/src/__tests__/integration/happy/client-engine/test.ts [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+  provider = "sqlite"
+  url      = "file:./dev.db"
+}
+
+generator client {
+  provider = "prisma-client-js"
+  ${engineType ? `engineType="${engineType}"` : ''}
+}
+
+// / User model comment
```
Comment (janpio) [inline packages/client/src/__tests__/integration/happy/client-engine/test.ts:21]: weird comment

File: scripts/bump-engines.ts [hunk 23]
```diff
@@ -20,7 +20,7 @@ async function main() {
   }
   await run(
     path.join(__dirname, '..'),
-    `pnpm update -r @prisma/engines-version@${version} @prisma/engines@${version}`,
+    `pnpm update -r @prisma/engines-version@${version} @prisma/engines@${version} @prisma/fetch-engine@${version} @prisma/get-platform@${version}`,
```
Comment (janpio) [inline scripts/bump-engines.ts:23]: huh @williamluke4?

**Review by janpio — APPROVED** [review]

File: scripts/bump-engines.ts (hunk 23 above)
Comment (Jolg42) [reply scripts/bump-engines.ts:23]: Note also done by Pierre in the interactive transaction PR

**Review by millsp — APPROVED** [review]
